src/pkcs11/misc.c

```diff
@@ -178,8 +178,7 @@ CK_RV push_login_state(struct sc_pkcs11_slot *slot,
 	struct sc_pkcs11_login *login = NULL;
 
 	if (!sc_pkcs11_conf.atomic || !slot) {
-		r = CKR_OK;
-		goto err;
+		return CKR_OK;
 	}
 
 	login = (struct sc_pkcs11_login *) malloc(sizeof *login);
```

Comment:
actually this should be `calloc` instead of `malloc` to avoid a bad free below in the error handling.